commit d53c0fede6e87cff80d0adf0782d1e73c99cbdb9
Author: Vadim Zeitlin <vadim@wxwidgets.org>
Date:   Wed May 2 12:52:09 2012 +0200

    Use real classes instead of #defines for deprecated wxFoobookEvents.
    
    Define wx{Choice,List,Note,Tool,Tree}bookEvent as real class instead of
    just #defining them as wxBookCtrlEvent to allow forward declaring them.
    
    This also requires all of the event macros related to them unfortunately but
    is probably still worth it to avoid breaking the compilation of existing code.
    
    Closes #14242.

diff --git a/include/wx/choicebk.h b/include/wx/choicebk.h
index 972a97c..2b60bf7 100644
--- a/include/wx/choicebk.h
+++ b/include/wx/choicebk.h
@@ -21,9 +21,6 @@
 
 class WXDLLIMPEXP_FWD_CORE wxChoice;
 
-wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_CHOICEBOOK_PAGE_CHANGED,  wxBookCtrlEvent );
-wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_CHOICEBOOK_PAGE_CHANGING, wxBookCtrlEvent );
-
 // wxChoicebook flags
 #define wxCHB_DEFAULT          wxBK_DEFAULT
 #define wxCHB_TOP              wxBK_TOP
@@ -110,16 +107,31 @@ private:
 // choicebook event class and related stuff
 // ----------------------------------------------------------------------------
 
-// wxChoicebookEvent is obsolete and defined for compatibility only
-#define wxChoicebookEvent wxBookCtrlEvent
-typedef wxBookCtrlEventFunction wxChoicebookEventFunction;
-#define wxChoicebookEventHandler(func) wxBookCtrlEventHandler(func)
+// wxChoicebookEvent is obsolete and defined for compatibility only, use
+// wxBookCtrlEvent directly in the new code.
+class wxChoicebookEvent : public wxBookCtrlEvent
+{
+public:
+    wxChoicebookEvent(wxEventType commandType = wxEVT_NULL, int winid = 0,
+                        int nSel = wxNOT_FOUND, int nOldSel = wxNOT_FOUND)
+        : wxBookCtrlEvent(commandType, winid, nSel, nOldSel)
+    {
+    }
+};
+
+typedef void (wxEvtHandler::*wxChoicebookEventFunction)(wxChoicebookEvent&);
+
+#define wxChoicebookEventHandler(func) \
+    wxEVENT_HANDLER_CAST(wxChoicebookEventFunction, func)
+
+wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_CHOICEBOOK_PAGE_CHANGED,  wxChoicebookEvent );
+wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_CHOICEBOOK_PAGE_CHANGING, wxChoicebookEvent );
 
 #define EVT_CHOICEBOOK_PAGE_CHANGED(winid, fn) \
-    wx__DECLARE_EVT1(wxEVT_COMMAND_CHOICEBOOK_PAGE_CHANGED, winid, wxBookCtrlEventHandler(fn))
+    wx__DECLARE_EVT1(wxEVT_COMMAND_CHOICEBOOK_PAGE_CHANGED, winid, wxChoicebookEventHandler(fn))
 
 #define EVT_CHOICEBOOK_PAGE_CHANGING(winid, fn) \
-    wx__DECLARE_EVT1(wxEVT_COMMAND_CHOICEBOOK_PAGE_CHANGING, winid, wxBookCtrlEventHandler(fn))
+    wx__DECLARE_EVT1(wxEVT_COMMAND_CHOICEBOOK_PAGE_CHANGING, winid, wxChoicebookEventHandler(fn))
 
 #endif // wxUSE_CHOICEBOOK
 
diff --git a/include/wx/listbook.h b/include/wx/listbook.h
index 48140dc..99c1f3e 100644
--- a/include/wx/listbook.h
+++ b/include/wx/listbook.h
@@ -108,18 +108,28 @@ private:
 // listbook event class and related stuff
 // ----------------------------------------------------------------------------
 
-// wxListbookEvent is obsolete and defined for compatibility only (notice that
-// we use #define and not typedef to also keep compatibility with the existing
-// code which forward declares it)
-#define wxListbookEvent wxBookCtrlEvent
-typedef wxBookCtrlEventFunction wxListbookEventFunction;
-#define wxListbookEventHandler(func) wxBookCtrlEventHandler(func)
+// wxListbookEvent is obsolete and defined for compatibility only, use
+// wxBookCtrlEvent directly in the new code.
+class wxListbookEvent : public wxBookCtrlEvent
+{
+public:
+    wxListbookEvent(wxEventType commandType = wxEVT_NULL, int winid = 0,
+                        int nSel = wxNOT_FOUND, int nOldSel = wxNOT_FOUND)
+        : wxBookCtrlEvent(commandType, winid, nSel, nOldSel)
+    {
+    }
+};
+
+typedef void (wxEvtHandler::*wxListbookEventFunction)(wxListbookEvent&);
+
+#define wxListbookEventHandler(func) \
+    wxEVENT_HANDLER_CAST(wxListbookEventFunction, func)
 
 #define EVT_LISTBOOK_PAGE_CHANGED(winid, fn) \
-    wx__DECLARE_EVT1(wxEVT_COMMAND_LISTBOOK_PAGE_CHANGED, winid, wxBookCtrlEventHandler(fn))
+    wx__DECLARE_EVT1(wxEVT_COMMAND_LISTBOOK_PAGE_CHANGED, winid, wxListbookEventHandler(fn))
 
 #define EVT_LISTBOOK_PAGE_CHANGING(winid, fn) \
-    wx__DECLARE_EVT1(wxEVT_COMMAND_LISTBOOK_PAGE_CHANGING, winid, wxBookCtrlEventHandler(fn))
+    wx__DECLARE_EVT1(wxEVT_COMMAND_LISTBOOK_PAGE_CHANGING, winid, wxListbookEventHandler(fn))
 
 #endif // wxUSE_LISTBOOK
 
diff --git a/include/wx/notebook.h b/include/wx/notebook.h
index 2311156..8524563 100644
--- a/include/wx/notebook.h
+++ b/include/wx/notebook.h
@@ -162,21 +162,31 @@ protected:
 // notebook event class and related stuff
 // ----------------------------------------------------------------------------
 
-// wxNotebookEvent is obsolete and defined for compatibility only (notice that
-// we use #define and not typedef to also keep compatibility with the existing
-// code which forward declares it)
-#define wxNotebookEvent wxBookCtrlEvent
-typedef wxBookCtrlEventFunction wxNotebookEventFunction;
-#define wxNotebookEventHandler(func) wxBookCtrlEventHandler(func)
+// wxNotebookEvent is obsolete and defined for compatibility only, use
+// wxBookCtrlEvent directly in the new code.
+class wxNotebookEvent : public wxBookCtrlEvent
+{
+public:
+    wxNotebookEvent(wxEventType commandType = wxEVT_NULL, int winid = 0,
+                        int nSel = wxNOT_FOUND, int nOldSel = wxNOT_FOUND)
+        : wxBookCtrlEvent(commandType, winid, nSel, nOldSel)
+    {
+    }
+};
+
+typedef void (wxEvtHandler::*wxNotebookEventFunction)(wxNotebookEvent&);
+
+#define wxNotebookEventHandler(func) \
+    wxEVENT_HANDLER_CAST(wxNotebookEventFunction, func)
 
-wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_NOTEBOOK_PAGE_CHANGED, wxBookCtrlEvent );
-wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_NOTEBOOK_PAGE_CHANGING, wxBookCtrlEvent );
+wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_NOTEBOOK_PAGE_CHANGED, wxNotebookEvent );
+wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_NOTEBOOK_PAGE_CHANGING, wxNotebookEvent );
 
 #define EVT_NOTEBOOK_PAGE_CHANGED(winid, fn) \
-    wx__DECLARE_EVT1(wxEVT_COMMAND_NOTEBOOK_PAGE_CHANGED, winid, wxBookCtrlEventHandler(fn))
+    wx__DECLARE_EVT1(wxEVT_COMMAND_NOTEBOOK_PAGE_CHANGED, winid, wxNotebookEventHandler(fn))
 
 #define EVT_NOTEBOOK_PAGE_CHANGING(winid, fn) \
-    wx__DECLARE_EVT1(wxEVT_COMMAND_NOTEBOOK_PAGE_CHANGING, winid, wxBookCtrlEventHandler(fn))
+    wx__DECLARE_EVT1(wxEVT_COMMAND_NOTEBOOK_PAGE_CHANGING, winid, wxNotebookEventHandler(fn))
 
 // ----------------------------------------------------------------------------
 // wxNotebook class itself
diff --git a/include/wx/toolbook.h b/include/wx/toolbook.h
index f3e3e3a..0f01ab6 100644
--- a/include/wx/toolbook.h
+++ b/include/wx/toolbook.h
@@ -21,10 +21,6 @@
 class WXDLLIMPEXP_FWD_CORE wxToolBarBase;
 class WXDLLIMPEXP_FWD_CORE wxCommandEvent;
 
-wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_TOOLBOOK_PAGE_CHANGED,  wxBookCtrlEvent );
-wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_TOOLBOOK_PAGE_CHANGING, wxBookCtrlEvent );
-
-
 // Use wxButtonToolBar
 #define wxTBK_BUTTONBAR            0x0100
 
@@ -127,17 +123,31 @@ private:
 // listbook event class and related stuff
 // ----------------------------------------------------------------------------
 
-// wxToolbookEvent is obsolete and defined for compatibility only
-#define wxToolbookEvent wxBookCtrlEvent
-typedef wxBookCtrlEventFunction wxToolbookEventFunction;
-#define wxToolbookEventHandler(func) wxBookCtrlEventHandler(func)
+// wxToolbookEvent is obsolete and defined for compatibility only, use
+// wxBookCtrlEvent directly in the new code.
+class wxToolbookEvent : public wxBookCtrlEvent
+{
+public:
+    wxToolbookEvent(wxEventType commandType = wxEVT_NULL, int winid = 0,
+                        int nSel = wxNOT_FOUND, int nOldSel = wxNOT_FOUND)
+        : wxBookCtrlEvent(commandType, winid, nSel, nOldSel)
+    {
+    }
+};
+
+typedef void (wxEvtHandler::*wxToolbookEventFunction)(wxToolbookEvent&);
+
+#define wxToolbookEventHandler(func) \
+    wxEVENT_HANDLER_CAST(wxToolbookEventFunction, func)
 
+wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_TOOLBOOK_PAGE_CHANGED,  wxToolbookEvent );
+wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_TOOLBOOK_PAGE_CHANGING, wxToolbookEvent );
 
 #define EVT_TOOLBOOK_PAGE_CHANGED(winid, fn) \
-    wx__DECLARE_EVT1(wxEVT_COMMAND_TOOLBOOK_PAGE_CHANGED, winid, wxBookCtrlEventHandler(fn))
+    wx__DECLARE_EVT1(wxEVT_COMMAND_TOOLBOOK_PAGE_CHANGED, winid, wxToolbookEventHandler(fn))
 
 #define EVT_TOOLBOOK_PAGE_CHANGING(winid, fn) \
-    wx__DECLARE_EVT1(wxEVT_COMMAND_TOOLBOOK_PAGE_CHANGING, winid, wxBookCtrlEventHandler(fn))
+    wx__DECLARE_EVT1(wxEVT_COMMAND_TOOLBOOK_PAGE_CHANGING, winid, wxToolbookEventHandler(fn))
 
 #endif // wxUSE_TOOLBOOK
 
diff --git a/include/wx/treebook.h b/include/wx/treebook.h
index a6bd61d..158efdb 100644
--- a/include/wx/treebook.h
+++ b/include/wx/treebook.h
@@ -228,28 +228,39 @@ private:
 // treebook event class and related stuff
 // ----------------------------------------------------------------------------
 
-// wxTreebookEvent is obsolete and defined for compatibility only
-#define wxTreebookEvent wxBookCtrlEvent
-typedef wxBookCtrlEventFunction wxTreebookEventFunction;
-#define wxTreebookEventHandler(func) wxBookCtrlEventHandler(func)
+// wxTreebookEvent is obsolete and defined for compatibility only, use
+// wxBookCtrlEvent directly in the new code.
+class wxTreebookEvent : public wxBookCtrlEvent
+{
+public:
+    wxTreebookEvent(wxEventType commandType = wxEVT_NULL, int winid = 0,
+                        int nSel = wxNOT_FOUND, int nOldSel = wxNOT_FOUND)
+        : wxBookCtrlEvent(commandType, winid, nSel, nOldSel)
+    {
+    }
+};
+
+typedef void (wxEvtHandler::*wxTreebookEventFunction)(wxTreebookEvent&);
+#define wxTreebookEventHandler(func) \
+    wxEVENT_HANDLER_CAST(wxTreebookEventFunction, func)
 
 
-wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_TREEBOOK_PAGE_CHANGED, wxBookCtrlEvent );
-wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_TREEBOOK_PAGE_CHANGING, wxBookCtrlEvent );
-wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_TREEBOOK_NODE_COLLAPSED, wxBookCtrlEvent );
-wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_TREEBOOK_NODE_EXPANDED, wxBookCtrlEvent );
+wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_TREEBOOK_PAGE_CHANGED, wxTreebookEvent );
+wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_TREEBOOK_PAGE_CHANGING, wxTreebookEvent );
+wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_TREEBOOK_NODE_COLLAPSED, wxTreebookEvent );
+wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COMMAND_TREEBOOK_NODE_EXPANDED, wxTreebookEvent );
 
 #define EVT_TREEBOOK_PAGE_CHANGED(winid, fn) \
-    wx__DECLARE_EVT1(wxEVT_COMMAND_TREEBOOK_PAGE_CHANGED, winid, wxBookCtrlEventHandler(fn))
+    wx__DECLARE_EVT1(wxEVT_COMMAND_TREEBOOK_PAGE_CHANGED, winid, wxTreebookEventHandler(fn))
 
 #define EVT_TREEBOOK_PAGE_CHANGING(winid, fn) \
-    wx__DECLARE_EVT1(wxEVT_COMMAND_TREEBOOK_PAGE_CHANGING, winid, wxBookCtrlEventHandler(fn))
+    wx__DECLARE_EVT1(wxEVT_COMMAND_TREEBOOK_PAGE_CHANGING, winid, wxTreebookEventHandler(fn))
 
 #define EVT_TREEBOOK_NODE_COLLAPSED(winid, fn) \
-    wx__DECLARE_EVT1(wxEVT_COMMAND_TREEBOOK_NODE_COLLAPSED, winid, wxBookCtrlEventHandler(fn))
+    wx__DECLARE_EVT1(wxEVT_COMMAND_TREEBOOK_NODE_COLLAPSED, winid, wxTreebookEventHandler(fn))
 
 #define EVT_TREEBOOK_NODE_EXPANDED(winid, fn) \
-    wx__DECLARE_EVT1(wxEVT_COMMAND_TREEBOOK_NODE_EXPANDED, winid, wxBookCtrlEventHandler(fn))
+    wx__DECLARE_EVT1(wxEVT_COMMAND_TREEBOOK_NODE_EXPANDED, winid, wxTreebookEventHandler(fn))
 
 
 #endif // wxUSE_TREEBOOK
diff --git a/src/common/nbkbase.cpp b/src/common/nbkbase.cpp
index a0b6a8b..5c2dabe 100644
--- a/src/common/nbkbase.cpp
+++ b/src/common/nbkbase.cpp
@@ -37,8 +37,8 @@
 
 extern WXDLLEXPORT_DATA(const char) wxNotebookNameStr[] = "notebook";
 
-wxDEFINE_EVENT( wxEVT_COMMAND_NOTEBOOK_PAGE_CHANGED, wxBookCtrlEvent );
-wxDEFINE_EVENT( wxEVT_COMMAND_NOTEBOOK_PAGE_CHANGING, wxBookCtrlEvent );
+wxDEFINE_EVENT( wxEVT_COMMAND_NOTEBOOK_PAGE_CHANGED, wxNotebookEvent );
+wxDEFINE_EVENT( wxEVT_COMMAND_NOTEBOOK_PAGE_CHANGING, wxNotebookEvent );
 
 // ----------------------------------------------------------------------------
 // XTI
diff --git a/src/generic/choicbkg.cpp b/src/generic/choicbkg.cpp
index 52fd088..5398569 100644
--- a/src/generic/choicbkg.cpp
+++ b/src/generic/choicbkg.cpp
@@ -49,8 +49,8 @@
 
 IMPLEMENT_DYNAMIC_CLASS(wxChoicebook, wxBookCtrlBase)
 
-wxDEFINE_EVENT( wxEVT_COMMAND_CHOICEBOOK_PAGE_CHANGING, wxBookCtrlEvent );
-wxDEFINE_EVENT( wxEVT_COMMAND_CHOICEBOOK_PAGE_CHANGED,  wxBookCtrlEvent );
+wxDEFINE_EVENT( wxEVT_COMMAND_CHOICEBOOK_PAGE_CHANGING, wxChoicebookEvent );
+wxDEFINE_EVENT( wxEVT_COMMAND_CHOICEBOOK_PAGE_CHANGED,  wxChoicebookEvent );
 
 BEGIN_EVENT_TABLE(wxChoicebook, wxBookCtrlBase)
     EVT_CHOICE(wxID_ANY, wxChoicebook::OnChoiceSelected)
diff --git a/src/generic/toolbkg.cpp b/src/generic/toolbkg.cpp
index 8ba3ac9..164c01f 100644
--- a/src/generic/toolbkg.cpp
+++ b/src/generic/toolbkg.cpp
@@ -45,8 +45,8 @@
 
 IMPLEMENT_DYNAMIC_CLASS(wxToolbook, wxBookCtrlBase)
 
-wxDEFINE_EVENT( wxEVT_COMMAND_TOOLBOOK_PAGE_CHANGING, wxBookCtrlEvent );
-wxDEFINE_EVENT( wxEVT_COMMAND_TOOLBOOK_PAGE_CHANGED,  wxBookCtrlEvent );
+wxDEFINE_EVENT( wxEVT_COMMAND_TOOLBOOK_PAGE_CHANGING, wxToolbookEvent );
+wxDEFINE_EVENT( wxEVT_COMMAND_TOOLBOOK_PAGE_CHANGED,  wxToolbookEvent );
 
 BEGIN_EVENT_TABLE(wxToolbook, wxBookCtrlBase)
     EVT_SIZE(wxToolbook::OnSize)
diff --git a/src/generic/treebkg.cpp b/src/generic/treebkg.cpp
index 140829a..b5e8f94 100644
--- a/src/generic/treebkg.cpp
+++ b/src/generic/treebkg.cpp
@@ -47,10 +47,10 @@
 
 IMPLEMENT_DYNAMIC_CLASS(wxTreebook, wxBookCtrlBase)
 
-wxDEFINE_EVENT( wxEVT_COMMAND_TREEBOOK_PAGE_CHANGING,  wxBookCtrlEvent );
-wxDEFINE_EVENT( wxEVT_COMMAND_TREEBOOK_PAGE_CHANGED,   wxBookCtrlEvent );
-wxDEFINE_EVENT( wxEVT_COMMAND_TREEBOOK_NODE_COLLAPSED, wxBookCtrlEvent );
-wxDEFINE_EVENT( wxEVT_COMMAND_TREEBOOK_NODE_EXPANDED,  wxBookCtrlEvent );
+wxDEFINE_EVENT( wxEVT_COMMAND_TREEBOOK_PAGE_CHANGING,  wxTreebookEvent );
+wxDEFINE_EVENT( wxEVT_COMMAND_TREEBOOK_PAGE_CHANGED,   wxTreebookEvent );
+wxDEFINE_EVENT( wxEVT_COMMAND_TREEBOOK_NODE_COLLAPSED, wxTreebookEvent );
+wxDEFINE_EVENT( wxEVT_COMMAND_TREEBOOK_NODE_EXPANDED,  wxTreebookEvent );
 
 BEGIN_EVENT_TABLE(wxTreebook, wxBookCtrlBase)
     EVT_TREE_SEL_CHANGED   (wxID_ANY, wxTreebook::OnTreeSelectionChange)
